core/services/keeper/validate.go

```diff
@@ -1,34 +1,79 @@
 package keeper
 
 import (
+	"reflect"
+
 	"github.com/pelletier/go-toml"
 	"github.com/pkg/errors"
 	uuid "github.com/satori/go.uuid"
 
+	"github.com/smartcontractkit/chainlink/core/logger"
 	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
 )
 
+// expectedObservationSourceRaw this is the expected observation source of the keeper job.
+const expectedObservationSourceRaw = `
+encode_check_upkeep_tx   [type=ethabiencode
+                          abi="checkUpkeep(uint256 id, address from)"
+                          data="{\"id\":$(jobSpec.upkeepID),\"from\":$(jobSpec.fromAddress)}"]
+check_upkeep_tx          [type=ethcall
+                          failEarly=true
+                          extractRevertReason=true
+                          contract="$(jobSpec.contractAddress)"
+                          gas="$(jobSpec.checkUpkeepGasLimit)"
+                          data="$(encode_check_upkeep_tx)"]
+decode_check_upkeep_tx   [type=ethabidecode
+                          abi="bytes memory performData, uint256 maxLinkPayment, uint256 gasLimit, uint256 adjustedGasWei, uint256 linkEth"]
+encode_perform_upkeep_tx [type=ethabiencode
+                          abi="performUpkeep(uint256 id, bytes calldata performData)"
+                          data="{\"id\": $(jobSpec.upkeepID),\"performData\":$(decode_check_upkeep_tx.performData)}"]
+perform_upkeep_tx        [type=ethtx
```

Comment:
@archseer remind me, if the app crashes after completing this eth_tx task, do we send a repeat transaction on the subsequent run?

Reply:
No, we made it idempotent: it will try to insert again but then find an existing eth_txes row and just return it instead:

https://github.com/smartcontractkit/chainlink/blob/e35f15f36bb9a07542fc67251aea17366358df64/core/services/bulletprooftxmanager/bulletprooftxmanager.go#L296-L299